controller/src/main/java/io/pravega/controller/task/KeyValueTable/TableMetadataTasks.java

```diff
@@ -0,0 +1,271 @@
+                                        return CompletableFuture.completedFuture(CreateKeyValueTableStatus.Status.SCOPE_NOT_FOUND);
+                                    }
+                                }), executor);
+                    } else {
+                       return CompletableFuture.completedFuture(CreateKeyValueTableStatus.Status.TABLE_EXISTS);
+                    }
+                }), e -> Exceptions.unwrap(e) instanceof RetryableException, Integer.MAX_VALUE, executor);
+    }
+
+    private CompletableFuture<Void> checkDone(Supplier<CompletableFuture<Boolean>> condition) {
+        return checkDone(condition, 100L);
+    }
+
+    private CompletableFuture<Void> checkDone(Supplier<CompletableFuture<Boolean>> condition, long delay) {
+        AtomicBoolean isDone = new AtomicBoolean(false);
+        return Futures.loop(() -> !isDone.get(),
+                () -> Futures.delayedFuture(condition, delay, executor)
+                        .thenAccept(isDone::set), executor);
+    }
+
+    private CompletableFuture<Boolean> isCreated(String scope, String kvtName, KeyValueTableConfiguration kvtConfig, Executor executor) {
+       return Futures.exceptionallyExpecting(kvtMetadataStore.getState(scope, kvtName, true, null, executor),
+                e -> Exceptions.unwrap(e) instanceof StoreException.DataNotFoundException, KVTableState.UNKNOWN)
+                .thenCompose(state -> CompletableFuture.completedFuture(state.equals(KVTableState.ACTIVE)));
+    }
+
+    @VisibleForTesting
+    public void setRequestEventWriter(EventStreamWriter<ControllerEvent> requestEventWriter) {
+        requestEventWriterRef.set(requestEventWriter);
+        writerInitFuture.complete(null);
+    }
+
+    /**
+     * This method takes an event and a future supplier and guarantees that if future supplier has been executed then event will
+     * be posted in request stream. It does it by following approach:
+     * 1. it first adds the index for the event to be posted to the current host.
+     * 2. it then invokes future.
+     * 3. it then posts event.
+     * 4. removes the index.
+     *
+     * If controller fails after step 2, a replacement controller will failover all indexes and {@link RequestSweeper} will
+     * post events for any index that is found.
+     *
+     * Upon failover, an index can be found if failure occurred in any step before 3. It is safe to post duplicate events
+     * because event processing is idempotent. It is also safe to post event even if step 2 was not performed because the
+     * event will be ignored by the processor after a while.
+     *
+     * @param event      Event to publish.
+     * @param futureSupplier  Supplier future to execute before submitting event.
+     * @return CompletableFuture<T> returned by Supplier or Exception.
+     */
+    @VisibleForTesting
+    <T> CompletableFuture<T> addIndexAndSubmitTask(ControllerEvent event, Supplier<CompletableFuture<T>> futureSupplier) {
```

Comment:
we should move this index and submit task to a helper class as this is identical to the code in streammetadatatasks.java

Reply:
done.